### ISSUE TITLE
ICE when using a generator borrowing data with feature(nll)

### ISSUE THREAD

**Issue by daboross** [rust-lang/rust#47055]
Hi! Ran into this earlier today when trying out `#![feature(nll)]` on a program which already uses generators.

Simplest test case:

```rust
#![crate_type = "lib"]
#![feature(conservative_impl_trait, generators, generator_trait, nll)]
use std::ops::Generator;

pub fn render<'a>(input: &'a str) -> impl Generator<Yield = (), Return = ()> + 'a {
    move || {
        let _t = input;
        yield ();
    }
}
```

Error:

```
error: internal compiler error: /checkout/src/librustc_mir/borrow_check/nll/universal_regions.rs:805: cannot convert `ReFree(DefId(0/0:252 ~ screeps_rs_ui[35dd]::rendering[0]::map_view[0]::render[0]), BrNamed(crate0:DefIndex(1:69), 'a))` to a region vid

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.24.0-nightly (77e189cd7 2017-12-28) running on x86_64-unknown-linux-gnu
```

Using `rustc 1.24.0-nightly (77e189cd7 2017-12-28)`

Reproduction on playpen: https://play.rust-lang.org/?gist=8973ce56db9aa1731173889ea012b8d9&version=nightly

**Comment by nikomatsakis** [comment]
Thanks =) Generators were not really tested, so I'm not too surprised.

**Comment by pnkfelix** [comment]
I think we even observe this locally via the `src/test/run-pass/generator/smoke.rs` test, which for me yields the following under `-Z nll`:

```
error: internal compiler error: librustc_mir/borrow_check/nll/universal_regions.rs:805: cannot convert `ReScope(Node(ItemLocalId(16)))` to a region vid
```

**Comment by nikomatsakis** [comment]
This works with https://github.com/rust-lang/rust/pull/47353 I believe

**Comment by bobtwinkles** [comment]
This is fixed in 2018-01-27 nightly and commit 7d6e5b9da0:
```none
┌ 0 ─ bob_twinkles@cheesecake ─ ~/Code/rust/rust-issue-47055
└╼ 1 rust-issue-47055 >> rustup toolchain link issue-47055 ./build/x86_64-unknown-linux-gnu/stage2
┌ 0 ─ bob_twinkles@cheesecake ─ ~/Code/rust/rust-issue-47055
└╼ rust-issue-47055 >> rustup override set issue-47055
info: override toolchain for '/home/bob_twinkles/Code/rust/rust-issue-47055' set to 'issue-47055'
┌ 0 ─ bob_twinkles@cheesecake ─ ~/Code/rust/rust-issue-47055
└╼ rust-issue-47055 >> cat << EOF > test-code.rs
> #![crate_type = "lib"]
> #![feature(conservative_impl_trait, generators, generator_trait, nll)]
> use std::ops::Generator;
> 
> pub fn render<'a>(input: &'a str) -> impl Generator<Yield = (), Return = ()> + 'a {
>     move || {
>         let _t = input;
>         yield ();
>     }
> }
> EOF
┌ 0 ─ bob_twinkles@cheesecake ─ ~/Code/rust/rust-issue-47055
└╼ rust-issue-47055 >> rustc ./test-code.rs 
┌ 0 ─ bob_twinkles@cheesecake ─ ~/Code/rust/rust-issue-47055
└╼ rust-issue-47055 >> ls
build  libtest_code.rlib  test-code.rs
┌ 0 ─ bob_twinkles@cheesecake ─ ~/Code/rust/rust-issue-47055
└╼ rust-issue-47055 >> file libtest_code.rlib 
libtest_code.rlib: current ar archive
```
The smoke test mentioned by @pnkfelix builds and runs as well, so I'm going to close this. Thanks for the report @daboross!
